cancer_ai/validator/rewarder.py

```diff
@@ -1,159 +1,193 @@
-
-        Args:
-            winner_hotkey: Competition winner's hotkey.
-            competition_id: Competition ID.
-            winner_model_result: Information about the winner's model.
-
-        """
-        result_improved = False
-        # Logic to check if new winner's model made any improvement. If not, keep current winner
-        if (
-            len(self.competition_leader_mapping) > 0
-            and competition_id in self.competition_leader_mapping
-        ):
-            current_leader_model_result = self.competition_leader_mapping[
-                competition_id
-            ].model_result
+class ModelScore(BaseModel):
+    date: datetime
+    score: float
 
-            result_improved = (
-                winner_model_result.score - current_leader_model_result.score > 0.001
-            )
 
-            if not result_improved:
-                winner_hotkey = self.competition_leader_mapping[competition_id].hotkey
+class CompetitionResultsStore(BaseModel):
+    # Structure: {competition_id: {hotkey: [ModelScore, ...]}}
+    score_map: dict[str, dict[Hotkey, list[ModelScore]]] = {}
+    # Structure: {competition_id: {hotkey: average_score}}
+    average_scores: dict[str, dict[Hotkey, float]] = {}
+    # Structure: {competition_id: (hotkey, score)}
+    current_top_hotkeys: dict[str, tuple[Hotkey, float]] = {}
+
+    def add_score(self, competition_id: str, hotkey: Hotkey, score: float, date: datetime = None):
+        """Add a score for a specific hotkey in a specific competition."""
+
+        # Initialize competition dictionaries if they don't exist
+        if competition_id not in self.score_map:
+            self.score_map[competition_id] = {}
+        if competition_id not in self.average_scores:
+            self.average_scores[competition_id] = {}
+
+        # Initialize hotkey list if it doesn't exist
+        if hotkey not in self.score_map[competition_id]:
+            self.score_map[competition_id][hotkey] = []
+
+        score_date = date if date is not None else datetime.now(timezone.utc)
+        
+        # Add the score
+        self.score_map[competition_id][hotkey].append(
+            ModelScore(date=score_date, score=score)
+        )
 
-        # reset the scores before updating them
-        self.scores = {}
+        # Sort by date and keep only the last HISTORY_LENGTH scores
+        self.score_map[competition_id][hotkey].sort(key=lambda x: x.date, reverse=True)
+        if len(self.score_map[competition_id][hotkey]) > HISTORY_LENGTH:
+            self.score_map[competition_id][hotkey] = self.score_map[competition_id][hotkey][-HISTORY_LENGTH:]
 
-        # get score and reduced share for the new winner
-        await self.get_miner_score_and_reduction(
-            competition_id, winner_hotkey, winner_model_result, result_improved
-        )
+        # Update the average score
+        self.update_average_score(competition_id, hotkey)
```

Comment:
try except catch here and there? is there possibility of some of these operations to fail?

Reply:
there is already try except inside this function